### PR TITLE
FFT implementaion

### PR DESCRIPTION
**Pull request by drslebedev**
This PR contains the following changes:
* Implementation of the iterative radix-2 FFT algorithm using bit-reversal permutation.
* FFT block contains 2 template parameters: data type and algorithm type.
* Add test of different FFT algorithms.
* Use `process_bulk` instead of `process_one`
* Allow user to set also output type
* Add additional meta information to the output `dataset`

### Clang-16
```text
┌────────────benchmark:─────────────┬──────┬──#N──┬─CTX-SW─┬───CPU cache misses────┬───CPU branch misses───┬─<CPU-I>─┬──min───┬──mean──┬─stddev─┬─median─┬──max───┬─total time─┬─ops/s─┐
│ std::complex<float> - fftw        │ PASS │ 100  │     0  │  29.6k / 179k = 16.5% │  6.76k / 355k =  1.9% │   92.1k │   6 us │  10 us │  37 us │   6 us │ 380 us │       1 ms │ 98.3k │
│ std::complex<float> - fft         │ PASS │ 100  │     1  │  8.18k / 148k =  5.5% │ 4.51k / 1.58M =  0.3% │    190k │  10 us │  11 us │   6 us │  10 us │  54 us │       1 ms │ 91.6k │
│ std::complex<float> - fftCT       │ PASS │ 100  │     0  │ 3.46k / 52.6k =  6.6% │ 7.80k / 18.8M =  0.0% │    1.0M │  75 us │  75 us │   1 us │  75 us │  85 us │       7 ms │ 13.4k │
│ std::complex<float> - magnitude   │ PASS │ 100  │     0  │  770  / 2.30k = 33.6% │   229  / 723k =  0.0% │   36.1k │   4 us │   4 us │ 208 ns │   4 us │   6 us │     399 us │  251k │
│ std::complex<float> - phase       │ PASS │ 100  │     0  │ 2.07k / 4.44k = 46.7% │ 3.57k / 3.04M =  0.1% │    146k │  12 us │  13 us │ 782 ns │  13 us │  18 us │       1 ms │ 78.7k │
│ std::complex<float> - dataset     │ PASS │ 100  │     0  │ 1.63k / 55.0k =  3.0% │ 4.09k / 1.52M =  0.3% │   67.5k │   4 us │   5 us │   1 us │   4 us │  17 us │     464 us │  215k │
├───────────────────────────────────┼──────┼──────┼────────┼───────────────────────┼───────────────────────┼─────────┼────────┼────────┼────────┼────────┼────────┼────────────┼───────┤
│ std::complex<double> - fftw       │ PASS │ 100  │     0  │  31.1k / 388k =  8.0% │  7.02k / 371k =  1.9% │   96.8k │   6 us │  10 us │  33 us │   6 us │ 338 us │     955 us │  105k │
│ std::complex<double> - fft        │ PASS │ 100  │     1  │  16.4k / 352k =  4.7% │ 4.52k / 1.63M =  0.3% │    212k │  11 us │  12 us │   7 us │  11 us │  80 us │       1 ms │ 85.4k │
│ std::complex<double> - fftCT      │ PASS │ 100  │     1  │  9.57k / 393k =  2.4% │ 16.6k / 22.1M =  0.1% │    1.5M │ 121 us │ 122 us │   2 us │ 122 us │ 140 us │      12 ms │  8.2k │
│ std::complex<double> - magnitude  │ PASS │ 100  │     0  │ 2.52k / 7.92k = 31.8% │ 1.35k / 1.28M =  0.1% │   68.5k │   7 us │   7 us │ 661 ns │   7 us │  13 us │     702 us │  142k │
│ std::complex<double> - phase      │ PASS │ 100  │     0  │ 1.35k / 38.3k =  3.5% │ 1.56k / 3.54M =  0.0% │    157k │  13 us │  13 us │ 262 ns │  13 us │  15 us │       1 ms │ 77.7k │
│ std::complex<double> - dataset    │ PASS │ 100  │     0  │  3.84k / 260k =  1.5% │ 3.39k / 1.53M =  0.2% │   71.3k │   5 us │   5 us │ 611 ns │   5 us │  11 us │     489 us │  204k │
├───────────────────────────────────┼──────┼──────┼────────┼───────────────────────┼───────────────────────┼─────────┼────────┼────────┼────────┼────────┼────────┼────────────┼───────┤
│ float - fftw                      │ PASS │ 100  │     0  │  29.5k / 142k = 20.7% │  15.0k / 959k =  1.6% │   85.8k │   2 us │   8 us │  53 us │   2 us │ 531 us │     763 us │  131k │
│ float - fft                       │ PASS │ 100  │     0  │  3.86k / 107k =  3.6% │ 3.54k / 1.65M =  0.2% │    193k │  10 us │  10 us │   2 us │  10 us │  29 us │       1 ms │ 99.7k │
│ float - magnitude                 │ PASS │ 100  │     0  │  514  / 1.60k = 32.0% │   203  / 364k =  0.1% │   18.2k │   2 us │   2 us │  53 ns │   2 us │   2 us │     187 us │  534k │
│ float - phase                     │ PASS │ 100  │     0  │  342  / 1.15k = 29.7% │ 1.60k / 1.53M =  0.1% │   72.8k │   6 us │   6 us │ 332 ns │   6 us │   9 us │     624 us │  160k │
│ float - dataset                   │ PASS │ 100  │     0  │ 1.19k / 9.24k = 12.9% │  2.24k / 928k =  0.2% │   41.1k │   3 us │   3 us │   1 us │   3 us │  12 us │     285 us │  351k │
├───────────────────────────────────┼──────┼──────┼────────┼───────────────────────┼───────────────────────┼─────────┼────────┼────────┼────────┼────────┼────────┼────────────┼───────┤
│ double - fftw                     │ PASS │ 100  │     1  │  31.3k / 251k = 12.5% │  15.1k / 969k =  1.6% │   87.6k │   3 us │   9 us │  60 us │   3 us │ 602 us │     872 us │  115k │
│ double - fft                      │ PASS │ 100  │     0  │  8.81k / 244k =  3.6% │ 3.93k / 1.59M =  0.2% │    206k │  10 us │  11 us │   4 us │  10 us │  55 us │       1 ms │ 91.9k │
│ double - magnitude                │ PASS │ 100  │     0  │  554  / 1.86k = 29.8% │   565  / 642k =  0.1% │   34.3k │   3 us │   3 us │ 106 ns │   3 us │   4 us │     349 us │  287k │
│ double - phase                    │ PASS │ 100  │     0  │  742  / 7.29k = 10.2% │  834  / 1.77M =  0.0% │   78.7k │   6 us │   6 us │ 137 ns │   6 us │   8 us │     644 us │  155k │
│ double - dataset                  │ PASS │ 100  │     1  │ 4.78k / 66.0k =  7.2% │  2.67k / 905k =  0.3% │   42.0k │   3 us │   3 us │   2 us │   3 us │  25 us │     308 us │  325k │
├───────────────────────────────────┼──────┼──────┼────────┼───────────────────────┼───────────────────────┼─────────┼────────┼────────┼────────┼────────┼────────┼────────────┼───────┤
└───────────────────────────────────┴──────┴──────┴────────┴───────────────────────┴───────────────────────┴─────────┴────────┴────────┴────────┴────────┴────────┴────────────┴───────┘
```

### gcc-13
```text
┌────────────benchmark:─────────────┬──────┬──#N──┬─CTX-SW─┬───CPU cache misses────┬───CPU branch misses───┬─<CPU-I>─┬──min──┬─mean──┬─stddev─┬─median─┬──max───┬─total time─┬─ops/s─┐
│ std::complex<float> - fftw        │ PASS │ 100  │     0  │  32.6k / 180k = 18.2% │  7.32k / 363k =  2.0% │   92.8k │  7 us │ 11 us │  42 us │   7 us │ 427 us │       1 ms │ 91.1k │
│ std::complex<float> - fft         │ PASS │ 100  │     0  │  6.23k / 136k =  4.6% │ 5.48k / 1.54M =  0.4% │    159k │ 35 us │ 36 us │   5 us │  35 us │  87 us │       4 ms │ 28.1k │
│ std::complex<float> - fftCT       │ PASS │ 100  │     1  │ 5.47k / 64.1k =  8.5% │ 6.22k / 14.8M =  0.0% │    966k │ 75 us │ 76 us │   3 us │  75 us │  95 us │       8 ms │ 13.2k │
│ std::complex<float> - magnitude   │ PASS │ 100  │     0  │  602  / 1.90k = 31.6% │   227  / 825k =  0.0% │   41.3k │  4 us │  4 us │  94 ns │   4 us │   5 us │     374 us │  267k │
│ std::complex<float> - phase       │ PASS │ 100  │     0  │ 1.18k / 2.68k = 44.2% │ 3.58k / 3.04M =  0.1% │    147k │ 12 us │ 12 us │ 981 ns │  12 us │  20 us │       1 ms │ 82.6k │
│ std::complex<float> - dataset     │ PASS │ 100  │     0  │ 1.77k / 46.3k =  3.8% │  2.61k / 992k =  0.3% │   46.3k │  7 us │  8 us │ 928 ns │   7 us │  17 us │     755 us │  132k │
├───────────────────────────────────┼──────┼──────┼────────┼───────────────────────┼───────────────────────┼─────────┼───────┼───────┼────────┼────────┼────────┼────────────┼───────┤
│ std::complex<double> - fftw       │ PASS │ 100  │     0  │  33.0k / 410k =  8.1% │  7.22k / 391k =  1.8% │   97.1k │  6 us │ 10 us │  30 us │   6 us │ 312 us │     970 us │  103k │
│ std::complex<double> - fft        │ PASS │ 100  │     0  │  10.4k / 361k =  2.9% │ 4.22k / 1.59M =  0.3% │    163k │ 37 us │ 37 us │   7 us │  37 us │ 106 us │       4 ms │ 26.8k │
│ std::complex<double> - fftCT      │ PASS │ 100  │     1  │  17.0k / 426k =  4.0% │ 9.24k / 17.8M =  0.1% │    1.3M │ 94 us │ 95 us │   3 us │  94 us │ 124 us │       9 ms │ 10.5k │
│ std::complex<double> - magnitude  │ PASS │ 100  │     0  │  694  / 4.95k = 14.0% │ 1.21k / 1.38M =  0.1% │   73.5k │  7 us │  7 us │ 120 ns │   7 us │   8 us │     728 us │  137k │
│ std::complex<double> - phase      │ PASS │ 100  │     0  │ 1.63k / 41.5k =  3.9% │ 1.56k / 3.54M =  0.0% │    158k │ 13 us │ 13 us │ 755 ns │  13 us │  18 us │       1 ms │ 77.0k │
│ std::complex<double> - dataset    │ PASS │ 100  │     0  │  5.00k / 288k =  1.7% │ 3.01k / 1.06M =  0.3% │   50.5k │  8 us │  9 us │ 742 ns │   9 us │  16 us │     938 us │  107k │
├───────────────────────────────────┼──────┼──────┼────────┼───────────────────────┼───────────────────────┼─────────┼───────┼───────┼────────┼────────┼────────┼────────────┼───────┤
│ float - fftw                      │ PASS │ 100  │     0  │  26.9k / 143k = 18.8% │  15.1k / 967k =  1.6% │   86.0k │  2 us │  8 us │  53 us │   2 us │ 532 us │     768 us │  130k │
│ float - fft                       │ PASS │ 100  │     1  │  5.31k / 104k =  5.1% │ 5.53k / 1.53M =  0.4% │    157k │ 34 us │ 35 us │   3 us │  34 us │  61 us │       3 ms │ 28.7k │
│ float - magnitude                 │ PASS │ 100  │     0  │  617  / 1.71k = 36.0% │   216  / 416k =  0.1% │   20.8k │  2 us │  2 us │  63 ns │   2 us │   2 us │     187 us │  535k │
│ float - phase                     │ PASS │ 100  │     1  │ 3.04k / 5.97k = 50.9% │ 2.03k / 1.53M =  0.1% │   73.5k │  6 us │  6 us │   2 us │   6 us │  26 us │     613 us │  163k │
│ float - dataset                   │ PASS │ 100  │     0  │ 1.30k / 9.60k = 13.5% │  1.40k / 703k =  0.2% │   31.8k │  4 us │  4 us │ 697 ns │   4 us │  11 us │     442 us │  226k │
├───────────────────────────────────┼──────┼──────┼────────┼───────────────────────┼───────────────────────┼─────────┼───────┼───────┼────────┼────────┼────────┼────────────┼───────┤
│ double - fftw                     │ PASS │ 100  │     0  │  27.7k / 263k = 10.5% │  15.7k / 991k =  1.6% │   88.4k │  3 us │  8 us │  53 us │   3 us │ 532 us │     811 us │  123k │
│ double - fft                      │ PASS │ 100  │     0  │  10.5k / 279k =  3.8% │ 4.21k / 1.59M =  0.3% │    159k │ 36 us │ 36 us │   4 us │  36 us │  77 us │       4 ms │ 27.5k │
│ double - magnitude                │ PASS │ 100  │     0  │  545  / 1.84k = 29.6% │   578  / 693k =  0.1% │   36.9k │  4 us │  4 us │ 125 ns │   4 us │   5 us │     369 us │  271k │
│ double - phase                    │ PASS │ 100  │     0  │  853  / 9.25k =  9.2% │  818  / 1.77M =  0.0% │   79.2k │  6 us │  6 us │ 131 ns │   6 us │   8 us │     644 us │  155k │
│ double - dataset                  │ PASS │ 100  │     0  │ 1.76k / 56.3k =  3.1% │  2.00k / 720k =  0.3% │   33.3k │  4 us │  5 us │ 496 ns │   4 us │   9 us │     453 us │  221k │
├───────────────────────────────────┼──────┼──────┼────────┼───────────────────────┼───────────────────────┼─────────┼───────┼───────┼────────┼────────┼────────┼────────────┼───────┤
└───────────────────────────────────┴──────┴──────┴────────┴───────────────────────┴───────────────────────┴─────────┴───────┴───────┴────────┴────────┴────────┴────────────┴───────┘
```

### CODE REVIEW

File: test/blocklib/algorithm/fft/fft_common.hpp
Comment (RalphSteinhagen) [inline test/blocklib/algorithm/fft/fft_common.hpp:1]: move this to algorithm since this is needed elsewhere as well.

File: test/blocklib/algorithm/fft/fft_common.hpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+template<ComplexType T, typename U>
+void
+computePhaseSpectrum(const std::vector<T> &fftOut, std::vector<U> &phaseSpectrum, bool outputInDb) {
+    // TODO: check vector sizes
+    using PrecisionType = typename T::value_type;
+    std::transform(fftOut.begin(), std::next(fftOut.begin(), phaseSpectrum.size()), phaseSpectrum.begin(), [outputInDb](const auto &c) {
+        const auto phase{ std::atan2(c.imag(), c.real()) };
+        return static_cast<U>(outputInDb ? PrecisionType(20.) * std::log10(std::abs(phase)) : phase);
```
Comment (RalphSteinhagen) [inline test/blocklib/algorithm/fft/fft_common.hpp:29]: Output in db doesn't make sense for phase... rather radians vs. degree. This should also include the (optional) unwrapping of the phase to have a linear definition.

... TODO should be handled. ;-)

File: test/blocklib/algorithm/fft/fft_common.hpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#define GRAPH_PROTOTYPE_ALGORITHM_FFT_COMMON_HPP
+
+#include "fft_types.hpp"
+#include <algorithm>
+#include <cmath>
+#include <vector>
+
+namespace gr::algorithm {
+
+template<ComplexType T, typename U>
```
Comment (RalphSteinhagen) [inline test/blocklib/algorithm/fft/fft_common.hpp:11]:
```cpp

template<typename T, typename U>
requires ComplexType<T> && std::floating_point<U>
void computeMagnitudeSpectrum(const std::vector<T> &fftOut, std::vector<U> &magnitudeSpectrum, std::size_t fftSize, bool outputInDb) {
    if (fftOut.size() != magnitudeSpectrum.size()) {
        throw std::invalid_argument("Input and output vector sizes must match");
    }

    std::ranges::transform(fftOut, magnitudeSpectrum.begin(), [fftSize, outputInDb](const auto &c) {
        const auto mag{ std::hypot(c.real(), c.imag()) * U(2.0) / static_cast<U>(fftSize) };
        return outputInDb ? U(20.0) * std::log10(mag) : mag;
    });
}
```

File: test/blocklib/algorithm/fft/fft_types.hpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
+#ifndef GRAPH_PROTOTYPE_ALGORITHM_FFT_TYPES_HPP
+#define GRAPH_PROTOTYPE_ALGORITHM_FFT_TYPES_HPP
+
+#include "node.hpp"
```
Comment (RalphSteinhagen) [inline test/blocklib/algorithm/fft/fft_types.hpp:4]: IMO this is better inlined where this is specifically needed to keep the logic and reasoning of the code local.

File: test/blocklib/core/fft/fft.hpp [hunk 4]
```diff
@@ -1,153 +1,63 @@
...
-concept LongDoubleType = std::is_same_v<T, std::complex<long double>> || std::is_same_v<T, long double>;
-
-template<typename T>
-concept FFTSupportedTypes = ComplexType<T> || std::integral<T> || std::floating_point<T>;
-
-inline static std::mutex fftw_plan_mutex;
-
-template<FFTSupportedTypes T>
-struct fft : node<fft<T>> {
+template<typename T, typename TImpl = gr::algorithm::FFT<typename FFTInternal<T>::InDataType>>
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:20]: nomen est omen: `TImpl` -> e.g. `FourierAlgoImpl`?

File: test/blocklib/core/fft/fft.hpp [hunk 5]
```diff
@@ -156,15 +66,15 @@ struct fft : node<fft<T>> {
 
     constexpr DataSet<U>
     process_one(T input) noexcept {
-        if constexpr (std::is_same_v<T, InHistoryType>) {
+        if constexpr (std::is_same_v<T, InDataType>) {
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:69]: I think this is a good if not __the__ use case for using process_bulk() and the input spans.... The history buffer would be redundant with that.

File: test/blocklib/core/fft/fft.hpp [hunk 6]
```diff
@@ -179,160 +89,105 @@ struct fft : node<fft<T>> {
...
+                    inData[i] *= windowVector[i];
                 }
             }
         }
     }
 
     void
-    computeFft() {
-        fftw<T>::execute(fftwPlan.get());
+    computeFFT() {
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:143]: let's inline this single line function... keeps the code more local and easier to reason about.

File: test/blocklib/core/fft/fft.hpp [hunk 7]
```diff
@@ -179,160 +89,105 @@ struct fft : node<fft<T>> {
...
+        // fftImpl.computeFFT(inData, outData);
+        outData = fftImpl.computeFFT(inData);
     }
 
     void
     computeMagnitudeSpectrum() {
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:149]: inline code -> simplifies local reasoning

File: test/blocklib/core/fft/fft.hpp [hunk 8]
```diff
@@ -179,160 +89,105 @@ struct fft : node<fft<T>> {
...
-        for (std::size_t i = 0; i < magnitudeSpectrum.size(); i++) {
-            const PrecisionType mag{ std::hypot(fftwOut[i][0], fftwOut[i][1]) * static_cast<PrecisionType>(2.0) / static_cast<PrecisionType>(fftSize) };
-            magnitudeSpectrum[i] = static_cast<U>(outputInDb ? static_cast<PrecisionType>(20.) * std::log10(std::abs(mag)) : mag);
-        }
+        gr::algorithm::computeMagnitudeSpectrum(outData, magnitudeSpectrum, fftSize, outputInDb);
     }
 
     void
     computePhaseSpectrum() {
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:154]: inline code -> simplifies local reasoning

File: test/blocklib/core/fft/fft.hpp [hunk 9]
```diff
@@ -1,153 +1,63 @@
 #ifndef GRAPH_PROTOTYPE_FFT_HPP
 #define GRAPH_PROTOTYPE_FFT_HPP
 
+#include "../../algorithm/fft/fft.hpp"
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:4]: Hardcoding the relative dependency isn't great. The directories should be included into the default cmake (& meson)build includes ... makes it easier to refactor/shift later.

File: test/blocklib/core/fft/fft.hpp [hunk 10]
```diff
@@ -179,160 +89,105 @@ struct fft : node<fft<T>> {
...
-            importWisdom();
-            fftwPlan = PlanUniquePtr(fftw<T>::plan(static_cast<int>(fftSize), fftwIn.get(), fftwOut.get(), sign, flags));
-            exportWisdom();
-        }
 
         initWindowFunction();
     }
 
     void
     initWindowFunction() {
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:176]: inline code -> simplifies local reasoning

File: test/blocklib/core/fft/fft.hpp [hunk 11]
```diff
@@ -1,153 +1,63 @@
...
+    PortIn<T>                                                               in;
+    PortOut<DataSet<U>>                                                     out;
+
+    TImpl                                                                   fftImpl;
+    Annotated<std::size_t, "FFT size", Doc<"FFT size">>                     fftSize{ 1024 };
+    Annotated<int, "window function", Doc<"window (apodization) function">> window{ static_cast<int>(WindowFunction::Hann) };
+    Annotated<float, "sample rate", Doc<"signal sample rate">, Unit<"Hz">>  sampleRate = 1.f;
+    Annotated<bool, "output in dB", Doc<"calculate output in decibels">>    outputInDb{ false };
+    std::vector<PrecisionType>                                              windowVector;
+    history_buffer<InDataType>                                              inputHistory{ fftSize };
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:38]: not needed/inefficient -> propose to drop in favour of using `process_bulk(...)`

N.B. you need to set the min/max `RequiredSamples<>`of the input port to the required default fftSize (i.e. 1024) during the init process and change if the user changes the settings.

File: test/blocklib/core/fft/fft.hpp [hunk 12]
```diff
@@ -1,153 +1,63 @@
...
+    std::vector<InDataType>                                                 inData{};
+    std::vector<OutDataType>                                                outData{};
+    std::vector<U>                                                          magnitudeSpectrum{};
+    std::vector<U>                                                          phaseSpectrum{};
+
+    FFT() { initAll(); }
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:44]: default constructor is OK... if you override this you must also implement the `FFT( ... initialiser_list ...)` one otherwise the default init/settings change won't work. Otherwise I'd propose to keep the rule of Zero..

File: test/blocklib/core/fft/fft.hpp [hunk 13]
```diff
@@ -1,153 +1,63 @@
...
             = delete;
 
-    ~fft() { clearFftw(); }
+    ~FFT() { clear(); }
 
     void
     settings_changed(const property_map & /*old_settings*/, const property_map &newSettings) noexcept {
         if (newSettings.contains("fftSize") && fftSize != inputHistory.capacity()) {
-            inputHistory = history_buffer<InHistoryType>(fftSize);
+            inputHistory = history_buffer<InDataType>(fftSize);
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:60]: need to also adjust the min/max required sample size of the port.

File: test/blocklib/core/fft/fft.hpp [hunk 14]
```diff
@@ -156,15 +66,15 @@ struct fft : node<fft<T>> {
 
     constexpr DataSet<U>
     process_one(T input) noexcept {
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:68]: we need to refactor this signature also (later) w.r.t. when we just want to export the raw mag|phase|complex spectrums as std::vector<..> + adding the required meta tags.

File: test/blocklib/core/fft/fft.hpp [hunk 15]
```diff
@@ -1,153 +1,63 @@
...
+    using OutDataType   = typename FFTInternal<T>::OutDataType;
+    using PrecisionType = typename FFTInternal<T>::PrecisionType;
+
+    PortIn<T>                                                               in;
+    PortOut<DataSet<U>>                                                     out;
+
+    TImpl                                                                   fftImpl;
+    Annotated<std::size_t, "FFT size", Doc<"FFT size">>                     fftSize{ 1024 };
+    Annotated<int, "window function", Doc<"window (apodization) function">> window{ static_cast<int>(WindowFunction::Hann) };
+    Annotated<float, "sample rate", Doc<"signal sample rate">, Unit<"Hz">>  sampleRate = 1.f;
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:35]: default signalName , unit etc are missing.

File: test/blocklib/core/fft/fft.hpp [hunk 16]
```diff
@@ -179,160 +89,105 @@ struct fft : node<fft<T>> {
         ds.timestamp = 0;
         const std::size_t N{ magnitudeSpectrum.size() };
 
-        ds.axis_names   = { "time", "fft.real", "fft.imag", "magnitude", "phase" };
+        ds.axis_names   = { "frequency", "fft.real", "fft.imag", "magnitude", "phase" };
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:92]: as discussed: the names should be derived from the signal_name (see default tags) that is passed and modified through the blocks. To note .. the forwarded tag needs to modify the name, eg. "FFT('old name')" etc.

File: test/blocklib/core/fft/fft.hpp [hunk 17]
```diff
@@ -179,160 +89,105 @@ struct fft : node<fft<T>> {
...
         ds.axis_units   = { "u.a.", "u.a.", "u.a.", "u.a.", "rad" };
         ds.extents      = { 4, static_cast<int32_t>(N) };
         ds.layout       = fair::graph::layout_right{};
         ds.signal_names = { "fft.real", "fft.imag", "magnitude", "phase" };
```
Comment (RalphSteinhagen) [inline test/blocklib/core/fft/fft.hpp:96]: meant these names ... the axis naming above is ok.

**Review by RalphSteinhagen — APPROVED** [review]
Generally OK. We need to revisit this in order to move it to a final production code quality (documentation, remove redundancies, corner cases, std::vector<...> return, selectable mag/phase/complex-only ...etc).

File: test/blocklib/core/fft/window.hpp [hunk 18]
```diff
@@ -2,23 +2,22 @@
 #ifndef GRAPH_PROTOTYPE_WINDOW_HPP
 #define GRAPH_PROTOTYPE_WINDOW_HPP
 
+#include <algorithm>
 #include <cmath>
 #include <numbers>
 #include <ranges>
 #include <vector>
 
 namespace gr::blocks::fft {
```
[the remark this answers is not in the document]
Comment (drslebedev) [reply test/blocklib/core/fft/window.hpp:11]: I put `window` also to the `gr::algorithm` namespace same as the rest of the `fft` algorithms.

File: test/blocklib/algorithm/fft/fft.hpp [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+#include "fft_types.hpp"
+#include "node.hpp"
+#include <ranges>
+
+namespace gr::algorithm {
+
+template<typename T>
+    requires(ComplexType<T> || std::floating_point<T>)
+struct FFT {
+    using PrecisionType = typename FFTInternal<T>::PrecisionType;
```
[the remark this answers is not in the document]
Comment (drslebedev) [reply test/blocklib/algorithm/fft/fft.hpp:14]: I put most of the code related to the FFT types to the separate file because it is used for both `FFT` and `FFTw` algorithms to get the proper input, output, precision etc. types. Basically now the types are defined automatically. But even if we give possibility to define both input and output types to the user, we need to add error/compatibility checker which is basically how it is implemented now.

File: test/blocklib/algorithm/fft/fft_common.hpp
Comment (drslebedev) [reply test/blocklib/algorithm/fft/fft_common.hpp:1]: The `fft_common.hpp` file is already in the `algorithm` directory. Or did you mean something else?

File: test/blocklib/algorithm/fft/fft_common.hpp (hunk 2 above)
Comment (drslebedev) [reply test/blocklib/algorithm/fft/fft_common.hpp:11]: The utput type `U` can be any numeric type not only `floating_point`, at least as it is now. 
magnitudeSpectrum.size can be also less than fft size, for examples in case of real input.

File: test/blocklib/algorithm/fft/fft_types.hpp (hunk 3 above)
Comment (drslebedev) [reply test/blocklib/algorithm/fft/fft_types.hpp:4]: This types are used at least for both FFt algorithms and in FFT QA. Probably it can be simplified if we do not support some rare combinations of input and output but just complex numbers, floats and doubles.

File: test/blocklib/core/fft/fft.hpp (hunk 5 above)
Comment (drslebedev) [reply test/blocklib/core/fft/fft.hpp:69]: I agree, process_bulk() probably is a better option and does not require history buffer. Probably this would be the way to go.
